### ISSUE TITLE
Refactoring: Lift `LocalDatabase` into DataSource APIs

### ISSUE THREAD

**Issue by dantengsky** [datafuselabs/datafuse#1568]
**Summary** 

TL;DR:

Suggestion:

- Decouple `LocalDatabase` into
  - TableFactoryRegistry
  - TableFactory 
- Provides an embedded MetaStore
  For unit tests. in-memory or backed by embedded KV DB
- Keep all the db metadata in MetaStore (except for `system` like virtual db/tables)
-----------------------------------------

`LocalDatabase` is quite convenient for us to verifying new features, or doing performance benchmarks. 
- we can bring up a test env easily without `Store`, it is a gift for UT
- `LocalDatabase` provides the functionalities of customizing `Table`  by `plan.engine`,  e.g. reading CSV/parquet data from the local filesystem/memory, it almost looks like DataSource.

But there is always room for enhancement:

- `LocalDatabase` is treated specially in Catalog;  local dbs and tables belonging to a local database are registered/created locally. Different computation nodes may have different sets of local databases/tables.  IMHO, this is not what we want (temp/tx level tables/views are quite different things I think)
- For the `special treatment`,  the implementation of Catalog looks at least a little bit messy: -(
- Although nothing prevents us from reading data from S3 in a Table which belongs to a `LocalDatabase`, that might be somehow weird ( _but not unacceptable, after a re-thought_)

IMHO, ideally, DB metadata and the interpretation of metadata into `Table` could be separated more clearly:

- DB metadata should belong to MetaStore(no matter Remote Or Embedded)
- Catalog is a component that exposes structural DB metadata to `Query` 
  - with friendly methods
  - and Virtual Database/Tables, i.e. the `system` db and virtual tables belong to it (there may be more as we keep going)
   
Thus, I suggest (roughly)

- Decouple `LocalDatabase` into
  - TableFactoryRegistry
  - TableFactory 
- Provides an embedded MetaStore
  For unit tests. in-memory or backed by embedded KV DB
- Keep all the db metadata in MetaStore 
  - except for `system` like virtual db/tables, which are views of db meta and system state

Note: In clickhouse, one can create a database by engine, e.g. if we create a clickhouse database, which engine is `MySQL`, CK will retrieve table metadata from the specified MySQL db.  The above suggestion does NOT provide such abilities.

**Comment by BohuTANG** [comment]
After the embedded MetaStore implement, the namespace register center (issue https://github.com/datafuselabs/datafuse/issues/883) will be more simple, no need a key-value store in query(this is done in https://github.com/datafuselabs/datafuse/pull/1011 but a bit more complex).
cc: @zhang2014

**Comment by drmingdrmer** [comment]
> IMHO, ideally, DB metadata and the interpretation of metadata into `Table` could be separated more clearly:

I do agree bro.

> * Provides an embedded MetaStore
>   For unit tests. in-memory or backed by embedded KV DB
> * Keep all the db metadata in MetaStore

StateMachine could be just what you need.  

- It provides with open/create API;
- To modify the state, just feed it a raft-log entry. I.e., it is just a meta store without a distributed WAL(raft log).
- To read the state, there are key spaces equipped with KV like APIs. Such as `self.kvs().get(key)?`, which you must be already familiar with.

**Comment by dantengsky** [comment]
> > IMHO, ideally, DB metadata and the interpretation of metadata into `Table` could be separated more clearly:
> 
> I do agree bro.
> 
> > * Provides an embedded MetaStore
> >   For unit tests. in-memory or backed by embedded KV DB
> > * Keep all the db metadata in MetaStore
> 
> StateMachine could be just what you need.
> 
> * It provides with open/create API;
> * To modify the state, just feed it a raft-log entry. I.e., it is just a meta store without a distributed WAL(raft log).
> * To read the state, there are key spaces equipped with KV like APIs. Such as `self.kvs().get(key)?`, which you must be already familiar with.

Thanks bro, it would be great to re-use the code of StateMachine.

**Comment by BohuTANG** [comment]
BTW, `LocalDatabase` is not really for testing, it's mainly for system information, such as `system` database.
```
mysql> use system;
mysql> show tables;
+---------------+
| name          |
+---------------+
| clusters      |
| contributors  |
| databases     |
| functions     |
| numbers       |
| numbers_local |
| numbers_mt    |
| one           |
| processes     |
| settings      |
| tables        |
| tracing       |
+---------------+

```

**Comment by dantengsky** [comment]
perhaps I miss emphasizing, by local database, I mean `datasources/local`.  `datasources/system/` not included
